nsgcli/grok.py

```diff
@@ -0,0 +1,107 @@
+"""
+This module implements subset of NetSpyGlass CLI commands
+
+:copyright: (c) 2021 by Happy Gears, Inc
+:license: Apache2, see LICENSE for more details.
+
+"""
+
+from __future__ import print_function
+
+import getopt
+import json
+import sys
+
+import nsgcli.api
+import nsgcli.sub_command
+import nsgcli.system
+import nsgcli.response_formatter
+
+HELP = """
+Parse text line with grok patterns pre-defined within NSG server and one custom grok pattern
+
+parse --text <text> [--pattern <pattern>]
+
+Examples:
+
+    parse --text "<13>May 18 11:22:43 carrier sshd: SSHD_LOGIN_FAILED: Login failed for user 'root' from host
+    '1.1.1.1'"
+
+    parse --text "<13>May 18 11:22:43 carrier sshd: SSHD_LOGIN_FAILED: Login failed for user 'root' from host 
+    1.1.1.1" --pattern "Login failed for user '%{WORD:login_name}'"
+
+"""
+
+
+class GrokCommands(nsgcli.sub_command.SubCommand, object):
+    def __init__(self, base_url, token, net_id, time_format=nsgcli.response_formatter.TIME_FORMAT_MS, region=None):
+        super(GrokCommands, self).__init__(base_url, token, net_id)
+        self.system_commands = nsgcli.system.SystemCommands(
+            self.base_url, self.token, self.netid, time_format=time_format, region=region)
+        self.current_region = region
+        if region is None:
+            self.prompt = 'grok # '
+        else:
+            self.prompt = 'grok [' + self.current_region + '] # '
+
+    ##########################################################################################
+    def do_parse(self, _):
+        """
+        Parse text line with grok patterns pre-defined within NSG server and one custom grok pattern
+
+        parse --text <text> [--pattern <pattern>]
+
+        Examples:
+
+            parse --text "<13>May 18 11:22:43 carrier sshd: SSHD_LOGIN_FAILED: Login failed for user 'root' from host
+            '79.174.187.54'"
+
+            parse --text "<13>May 18 11:22:43 carrier sshd: SSHD_LOGIN_FAILED: Login failed for user
+            'root' from host 1.1.1.1" --pattern "Login failed for user '%{WORD:login_name}'"
+        """
+        try:
+            opts, args = getopt.getopt(sys.argv[3:],
+                                       't:p:',
+                                       ['text=', 'pattern='])
+        except getopt.GetoptError as ex:
+            print('UNKNOWN: Invalid Argument:' + str(ex))
+            raise InvalidArgsException
+
+        txt = None
+        pattern = None
+        for opt, arg in opts:
+            if opt in ['-t', '--text']:
+                txt = arg
+            elif opt in ('-p', '--pattern'):
+                pattern = arg
+
+        if not txt:
+            print('--text parameter is mandatory')
+            raise InvalidArgsException
+
+        request = 'v2/grok/parser'
+        try:
```

Comment:
I did not look at the nsg-grokapi side. Our API usually includes "net" match , as in "/v2/ui/net/1/status". The purpose is to preserve the option of having true multitenancy in our servers should we decide to implement it. If your API call path does not match "network id", then you lose this ability. It is easier to add support for network id match when you develop the service rather than to put it off for later, when we potentially may have to retrofit it into many services